### PR TITLE
multinets

### LINKED ISSUE #652: consider refactoring simulation calls in netsim and/or netdx
We have 11 in `net.mod.simnet.R` and 12 in `netdx.R`.  This seems like more than should be necessary, given the number of fundamentally distinct cases we need to handle.

In `simnet`, the `tergm` and `tergmLite` cases could be handled together by simply defining the inputs and outputs appropriately.  (For single network models, as in core EpiModel, network stats can be extracted during simulation in the `tergmLite` case.)

For `netdx`, this refactoring could be part of larger changes made for #490.

### LINKED ISSUE #689: Remove nodefactor from diss models in future release
This functionality was deprecated starting with the merger of the netdx_heterodiss branch into main 56811b9dc20e88da5af2ea719ed2bece84592e63

### LINKED ISSUE #758: Default of nwstats.formula in multi-layer networks
This is currently an issue in extension packages (HIV, COVID) only, but probably should be handled here. With multi-layer networks, it is necessary to explicitly define the `nwstats.formula` for each layer as:

```r
nwstats.formula.1 = ...
nwstats.formula.2 = ...
nwstats.formula.3 = ...
``` 

for each layer. Nearly always we would like to use the main ERGM formula used in the model, so we input `"formation"` on the right hand side. Could this default in single layer networks be repeated for each layer in multi-layer networks so that it is not necessary to explicitly define `nwstats.formula.X = "formation"`?

### LINKED ISSUE #759: consider removing `init_tergmLite`
The (few) essential bits of it could be absorbed into `initialize.net`.  This has already been done in the `simnet_updates` branch, but we will want to discuss those changes as a whole at a later date.

Note that `EpiModelHIV` will need to be updated accordingly.

### LINKED ISSUE #760: consider removing `sim_nets_t1`
IIRC this was originally introduced to facilitate the "list of networks" approach to simulating an `ergm` in `netsim` (which has since been replaced by `tergm` simulation with `networkDynamic`s).  Much of `sim_nets_t1` is redundant with `resim_nets`, and the initial `ergm` simulation (which is not redundant with `resim_nets`) could be moved into `initialize.net`.  Some small tweaks may be needed elsewhere to make this work.

Note that `sim_nets_t1` has already been removed in the `simnet_updates` branch, but we will want to discuss those changes as a whole at a later date.

An alternative to completely removing `sim_nets_t1` may be to refactor the `tergm` simulate calls so that more code can be shared between `sim_nets_t1` and `resim_nets`.

### LINKED ISSUE #764: is it allowed to initialize a netsim run from a netsim object when `start == 1`?
In `crosscheck.net` we appear to allow this in the following check:

https://github.com/EpiModel/EpiModel/blob/8755c1e2eca75c9990a1e8cb0b4588e37eeaedf6/R/net.inputs.R#L1066-L1072

but some of the code that follows in `crosscheck.net`, as well as code in `initialize.net`, appears to assume a `netest` object when `start == 1`.

### PR DESCRIPTION
**Pull request by chad-klumb**
This PR combines changes refactoring `simnet`, removing deprecated functionality (`nodefactor` dissolution and `stergm` controls), and adding basic support for multiple network models in `netsim`.

Currently a draft.

closes #652 (refactoring `simnet`)

closes #689 (removing `nodefactor` dissolution)

closes #758 (adding `multilayer` control specification)

closes #759 (removing `init_tergmLite`)

closes #760 (leaving a reworked `sim_nets_t1`)

closes #764 (disallowing `start == 1` initialization from a `netsim` object)

references #669, #757

### CODE REVIEW

**Comment by smjenness** [comment]
@chad-klumb : I have asked @AdrienLeGuillou to take a look at this work in progress, so he may have some running questions/comments we move forward.

**Comment by AdrienLeGuillou** [comment]
I have some concerns regarding the network controls, I fear that the current implementation could lead to potentially hard to debug issues as `get_control` silently returns the default value if `out` is not a "multilayer".

I think it may be worthwhile to ensure that all networks have their own controls in the `dat` object. That could be inside `dat$control$network$<a list per network>` in `dat$network_controls` or something else entirely. We could then have a `get_network_control` function where the network index is mandatory. 

Such a separation would make it clear that these controls are specific to networks and to one in particular. I know that the new user that I help (and myself at the begining) have a hard time understanding what does what in the arguments we pass to control. Having a stricter structure would help to clarify things a bit. 

Also, the networks have a very unique position in EpiModel as most user rely entirely on the fact that `statnet` will work as expected (and for good reason). Therefore, having the network controls as their own thing would make sense. (Again the specifics of where we store them is not my point).

Finally, we could provide utilities like the `multilayer` class to ease the setup.

what do you think @smjenness and @chad-klumb?

NB: you know by now how I prefer simple generic structures and higher level abstraction to manipulate them over flexible functions that can handle many different situations. I believe the former is way easier to debug and update but if you think it's not the case I will not be offended :-)

**Comment by chad-klumb** [comment]
Adrien, are you suggesting just a back-end change (of how we get, set, and store controls that are allowed to be specific to each network, when dealing with the `dat` object), or also a front-end change (of how those controls are specified by the user)?

I would note that making `get_control(dat, name, network = whatever)` return the full control value when `dat$control[[name]]` is not a `multilayer` was one way of addressing Sam's concern in #758 (and the corresponding issue for `set.control.*` arguments).  We can be more explicit in the back-end (at the cost of some extra code), converting the relevant controls to `multilayer` on initialization (if they are not passed in that way by the user), and then calling a more restrictive `get_network_control` accessor after that.  (We can also check that if the user does pass a `multilayer` argument then it has the correct length given the number of networks being simulated.)

**Comment by AdrienLeGuillou** [comment]
my preference would be backend change and accessor changes. For the end user (i.e. the one that simply runs the final model) I don't think it matters too much. But for the one implementing modules and specific things, I would prefer to have things well separated.

I believe having a more restrictive `get_network_control` (or any name) would make the code easier to understand especially when debugging. My preference is always to have more simple functions, especially in the API part. For the final user I understand the benefice of a simple interface with flags.

And I am completely on board with the `multilayer` class that allow to check the correctness of settings to be passed to `control.net`.

**Comment by chad-klumb** [comment]
updated along the lines we discussed... let me know if you have any more input!

**Comment by chad-klumb** [comment]
note that network size and vertex attributes continue to be drawn from the first network when multiple networks are used

I've updated the documentation to emphasize this

**Comment by AdrienLeGuillou** [comment]
thanks for the changes @chad-klumb. 

I have another proposition but I am unsure if it's the best way to go or if it is in the scope of this PR. So I am asking for your opinion (and maybe @smjenness 's) and not requesting anything:

What do you thing of having (even in mid / long term) a `dat$network` sub-list that would contain everything that is relevant to the networks? I see it as a list of size `number of networks` and for each network we would have its `control`, `el`, `nwparam`, etc.

The idea is to view each network as it's own sub object which would clearly identify what refers to it and what it impacts. 

Again, I understand that it can be a lot of code to change and that its not even desirable, so don't hesitate to disagree :-)

File: R/net.accessor.R [hunk 1]
```diff
@@ -438,6 +438,35 @@ get_control <- function(dat, item, override.null.error = FALSE) {
   return(out)
 }
 
+#' @rdname net-accessor
+#' @param network index of network for which to get control
+#' @export
+get_network_control <- function(dat, item, override.null.error = FALSE, network) {
```
Comment (AdrienLeGuillou) [inline R/net.accessor.R:444]: `network` could be the second argument for consistency with the other functions with a mandatory `network` argument

File: R/net.accessor.R [hunk 2]
```diff
@@ -438,6 +438,35 @@ get_control <- function(dat, item, override.null.error = FALSE) {
...
+  if (missing(network)) {
+    stop("`get_network_control` requires `network` argument.")
+  }
```
Comment (AdrienLeGuillou) [inline R/net.accessor.R:447]: `network` being a mandatory argument, we should not handle it with `missing` and instead let R crash in the `normal` way if it's not provided.

Also, it is usually agreed that `missing` should be avoid in general and replaced with a default value. Usually, `network = NULL` in the function signature and `if (is.null(network))` makes it immediately visible that the `network` argument is optional without the need to read the body of the function or the documentation.

File: R/net.mod.infection.R [hunk 3]
```diff
@@ -56,7 +56,7 @@ infection.net <- function(dat, at) {
   if (nElig > 0 && nElig < nActive) {
 
     # Get discordant edgelist
-    del <- discord_edgelist(dat, at, network = 1)
+    del <- do.call(rbind, lapply(seq_along(dat$nwparam), discord_edgelist, dat = dat, at = at, include.network = TRUE))
```
Comment (AdrienLeGuillou) [inline R/net.mod.infection.R:59]: now that `dplyr` is imported by `EpiModel`, we can use `dplyr::bind_rows` instead of `do.call(rbind, list_of_dataframe)`. It makes for an easier to read code and it's way faster.

the code could be:

```r
del_lists <- lapply(seq_along(dat$nwparam), discord_edgelist, dat = dat, at = at, include.network = TRUE)
del <- dplyr::bind_rows(del_lists)
```

File: R/net.mod.infection.R [hunk 4]
```diff
@@ -178,7 +178,7 @@ infection.2g.net <- function(dat, at) {
...
```
Comment (AdrienLeGuillou) [inline R/net.mod.infection.R:181]: same comment

File: R/net.mod.infection.R [hunk 5]
```diff
@@ -320,6 +322,10 @@ discord_edgelist <- function(dat, at, network = 1, infstat = "i") {
     }
   }
 
+  if (NROW(del) > 0 && include.network == TRUE) {
+    del <- cbind(del, network = network)
```
Comment (AdrienLeGuillou) [inline R/net.mod.infection.R:326]: `cbind` can be replaced with `dplyr::bind_cols` which is faster

File: R/net.accessor.R (hunk 2 above)
Comment (chad-klumb) [reply R/net.accessor.R:447]: I don't think R's default handling of a missing argument used as an index in a `list` is really what we want here.  For example, if we define

```
f <- function(a,b) a[[b]]
```

then

```
f(a = list(1,2,3))
```

evaluates to `NULL`; there is no error thrown, even though `b` is missing and is used to index `a`.

We can restore an error message without an explicit `missing` check by calling `force(b)` before returning `a[[b]]`.

Comment (chad-klumb) [reply R/net.accessor.R]: the `force` mechanism is used in dbdaab3972fab3d4a912d9ed89b844ebb50e0080 but I will leave this open for now in case there is more to discuss

Comment (AdrienLeGuillou) [reply R/net.accessor.R]: Oh I did not know this case! Thanks.

In this case I can see that `missing` is actually the right call, sorry for this comment

File: R/net.mod.infection.R (hunk 4 above)
Comment (chad-klumb) [reply R/net.mod.infection.R]: addressed in 17e2faf3895ebb01e215ec77b7351b204aa2d099

**Comment by chad-klumb** [comment]
In principle it might be worth considering some restructuring of the `dat` object, but I do think that is beyond the scope of this PR and should be addressed as a separate issue.

**Comment by chad-klumb** [comment]
I'll be giving this some final review over the weekend and will convert it to an open PR when I'm done with that.

**Comment by chad-klumb** [comment]
closes #669

closes #757

closes #767

**Comment by chad-klumb** [comment]
Added `multilayer` support for `tergmLite.track.duration` as you may wish to track duration for some networks and not others. 
 (Note that tracking duration for large or dense networks with short edge durations may be relatively costly, as `lasttoggle` will accumulate many changes.)  The default remains `FALSE` for all networks.

**Comment by chad-klumb** [comment]
Ready for testing/review.  Pairs with `EpiModel/EpiModelHIV-p@multinets`.

**Comment by chad-klumb** [comment]
There are example usages in

https://github.com/EpiModel/EpiModel/blob/multinets/tests/testthat/test-multinets.R

and

https://github.com/EpiModel/EpiModelHIV-p/blob/multinets/tests/testthat/test-model.R

File: R/net.inputs.R [hunk 6]
```diff
@@ -1033,9 +1049,13 @@ crosscheck.net <- function(x, param, init, control) {
...
 
       # Main class check ----------------------------------------------------
-      if (!inherits(x, c("netest", "netsim"))) {
-        stop("x must be either an object of class netest or class netsim",
-             call. = FALSE)
+      if (inherits(x, "netest")) {
+        x <- list(x)
+      }
+      if (!inherits(x, "list") || length(x) == 0 ||
+          !all(unlist(lapply(x, inherits, "netest")))) {
```
Comment (AdrienLeGuillou) [inline R/net.inputs.R:1056]: could we use 
`vapply(x, inherits, "netest", logical(1))` instead?

File: R/net.mod.init.R [hunk 7]
```diff
@@ -23,18 +27,13 @@ initialize.net <- function(x, param, init, control, s) {
...
-    dat$nwparam <- list()
-    dat$nwparam[[1]] <- x[!(names(x) %in% c("fit", "newnetwork"))]
-
-    # Initial Network Simulation ----------------------------------------------
-
-    if (get_control(dat, "resimulate.network") == TRUE) {
-      nsteps <- 1
-    } else {
-      nsteps <- get_control(dat, "nsteps")
+    if (is(x, "netest")) {
```
Comment (AdrienLeGuillou) [inline R/net.mod.init.R:30]: maybe we could use `inherits` here to future proof this?

**Comment by AdrienLeGuillou** [comment]
Impressive work! Apart from the small comment above all seems good to me. I will run some tests with the EpiModelHIV branch as well and get back to you

File: R/net.inputs.R (hunk 6 above)
Comment (chad-klumb) [reply R/net.inputs.R]: I've switched to `vapply` and `inherits` as suggested.

**Comment by smjenness** [comment]
@chad-klumb : could you summarize which statnet packages currently in the DESCRIPTION::remotes field need to get published on CRAN for this branch to be CRAN-ready? Thanks!

**Comment by chad-klumb** [comment]
requires `tergm@master` for the bug fix in statnet/tergm#99

**Comment by smjenness** [comment]
Thanks @chad-klumb! We can/will review these changes at our meeting upcoming on Thursday. 

One idea in advance: would you be willing/able to write up a short working example as an addition to the EpiModel Gallery (https://github.com/EpiModel/EpiModel-Gallery)? You can see the other examples there as templates. The general idea is to provide a little more details about the implementation than is available in the function docs and/or unit tests.

**Comment by chad-klumb** [comment]
I can construct a gallery example, probably after our Thursday meeting (there are examples in the tests and in `EpiModelHIV-p@multinets` that we can look at on Thursday).

**Comment by chad-klumb** [comment]
This should be ready for review/merge.

**Review by AdrienLeGuillou — COMMENTED** [review]
Well done @chad-klumb, my main comment is about the form of the `dat.updates` functions. 
I tested with EMHIV-p as well with the template and it works correctly. I will make a specific review for this code as well

File: R/net.inputs.R [hunk 8]
```diff
@@ -727,17 +730,32 @@ init.net <- function(i.num, r.num, i.num.g2, r.num.g2,
...
-#'        help file for [`netdx`] for details and examples on specifying this parameter.
-#' @param set.control.stergm Deprecated control argument of class `control.simulate.network`. Use
-#'        `set.control.tergm` instead.
+#'        help file for [`netdx`] for details and examples on specifying this parameter. Supports
+#'        [`multilayer`] specification.
 #' @param save.diss.stats If `TRUE`, `netsim` will compute and save duration and dissolution
 #'        statistics for plotting and printing, provided `save.network` is `TRUE`, `tergmLite` is
 #'        `FALSE`, and the dissolution model is homogeneous.
+#' @param dat.updates Either `NULL`, a single function taking arguments `dat`,
+#'        `at`, and `network`, or a list of functions of length one greater
```
[the remark this answers is not in the document]
Comment (chad-klumb) [reply R/net.inputs.R:744]: Currently `dat.updates` uses `at = 0L` to indicate the initial cross-sectional simulation, and `at = 1L` to indicate the first time step simulation (both taking place in `sim_nets_t1`). If we relied on `get_current_timestep` with the way things are currently set up, I think they would both come back as `1L` (so the user couldn't distinguish between the initial `ergm` simulation and the first timestep `tergm` simulation).

**Comment by AdrienLeGuillou** [comment]
@smjenness @chad-klumb 
just to let you know that I re-ran an EpiModelHIV-Template base project with this branch as well with the [companion EpiModelHIV-p one](https://github.com/EpiModel/EpiModel/pull/765) and it worked perfectly both for the local and HPC runs using slurmworkflows

**Comment by smjenness** [comment]
Mostly for my own notes here.... There are multiple layers of merges and publications necessary here.

1. Publish `networkLite` to CRAN. `networkLite@main` contains the merged EpiModel/networkLite#8.
2. Once `networkLite` on CRAN, EpiModel/EpiModel#828 should be updated to remove the Github remote, then merged.
3. Then EpiModel/EpiModel#824 can be merged (comments/review from @AdrienLeGuillou are requested)
4. Then EpiModel/EpiModel#765 can be merged (the main multinets branch)
5. Publish EpiModel to CRAN with this. Probably will go to v2.4.0 given the size of the changes.
6.

**Comment by chad-klumb** [comment]
It does look like EpiModelCOVID is out of date with some of the changes in `multinets` (and stacked PRs).  I'll open an issue to address that.

**Comment by chad-klumb** [comment]
The other merges you've mentioned look comprehensive to me (given the current state of things).

**Review by smjenness — APPROVED** [review]
